### PR TITLE
Add support to `switch` in `transpile`

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

For the most part, the heavy lifting for this was already done in the previous support for control flow.  This commit refactors some of the handling into larger, more descriptive helper functions, and most of the rest of the work is in testing.

I'm fairly confident there are some register-related problems to do with mappings in `DAGCircuit.substitute_node_*` and `QuantumCircuit.compose` right now, in both existing control flow and the new switch statement, which will need revisiting.

### Details and comments

~Depends on #9916, #9919, #9926.  This PR currently contains a cherry-pick of all three of those, which will be rebased out before merge.~

### CODE REVIEW

**Review by mtreinish — DISMISSED** [review]
I reviewed the contents of the second commit, it all looks good to me. Longer term I have some concerns about how general our interface for compiling arbitrary control flow operations. But as I read through the rest of the PR I realized/remembered that is a much larger amount of effort.

But aside from that the actual change here is pretty straightforward it's just adding a bit of edge case handling to the existing control flow handling code to extend it for the extra needs of switch. The test coverage looks like what I'd expect too, the big one for me is the full path `transpile()` tests with the recursive visitor validating the layout (which I remember caught issues back in 0.22).

Marking as approved as this LGTM in its current form, but obviously it needs to be rebased after #9926 merges. I'll re-approve when that happens.

File: qiskit/dagcircuit/dagcircuit.py
Comment (jakelishman) [reply qiskit/dagcircuit/dagcircuit.py:905]: Yeah, for sure this isn't super robust and some alternate form would probably be better - I think that would be tied to some better IR.  I'm not super convinced with the idea of supporting custom control flow, though - I think that might just be too much complexity with little gain for Terra.

File: test/python/compiler/test_transpiler.py
Comment (jakelishman) [reply test/python/compiler/test_transpiler.py:1786]: Yeah, I could improve this test now that we can round-trip.  Maybe a follow-up is clearer, so it doesn't coincide in the history with changes to the behaviour under test.

**Comment by jakelishman** [comment]
Now properly rebased over `main` and I fixed the mistaken variable name causing a test failure.

**Comment by jakelishman** [comment]
I've also fully fleshed out the release note for the `switch` statement now, since this is the last feature PR for it for this release.

File: releasenotes/notes/switch-case-9b6611d0603d36c0.yaml
Comment (mtreinish) [inline releasenotes/notes/switch-case-9b6611d0603d36c0.yaml:38]: I probably should have brought this up during the QASM 3 review (although it might have been in there but I don't remember), but are we documenting anywhere what the support policy is around experimental flags. Like what happens if/when the openqasm spec adopts the switch syntax but the final version diverges from what's in the experimental flag. Are we potentially going to be a situation where we have orphaned qasm files generated by qiskit which can no longer be loaded used by newer versions of anything? I think it's fine if that's the case but I'm thinking we should document this.

File: releasenotes/notes/switch-case-9b6611d0603d36c0.yaml
Comment (mtreinish) [inline releasenotes/notes/switch-case-9b6611d0603d36c0.yaml:36]:
```suggestion
      expect it to be adopted in the near future.  Qiskit Terra has experimental support for
      exporting this statement to the OpenQASM 3 syntax proposed in the linked pull request, using
```

File: releasenotes/notes/switch-case-9b6611d0603d36c0.yaml
Comment (mtreinish) [inline releasenotes/notes/switch-case-9b6611d0603d36c0.yaml:35]:
```suggestion
      design and consideration <https://github.com/openqasm/openqasm/pull/463>`__, which is
      expected to be adopted in the near future.  Qiskit Terra has experimental support for
```

Comment (jakelishman) [reply releasenotes/notes/switch-case-9b6611d0603d36c0.yaml:35]: Done in a3b5a780f - I couldn't apply the suggestions because they overlap.

File: releasenotes/notes/switch-case-9b6611d0603d36c0.yaml
Comment (jakelishman) [reply releasenotes/notes/switch-case-9b6611d0603d36c0.yaml]: Done in a3b5a780f.

File: releasenotes/notes/switch-case-9b6611d0603d36c0.yaml
Comment (jakelishman) [reply releasenotes/notes/switch-case-9b6611d0603d36c0.yaml]: That sounds like a good idea to flag for me.  I've made #9990 since it's a docs-only change that can land separately, in case there's more about the policy to discuss.

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for the quick updates

**Pull request merged**
